src/lib/expression/evaluation/expression_evaluator.cpp

```diff
@@ -253,27 +254,97 @@ ExpressionEvaluator::_evaluate_in_expression<ExpressionEvaluator::Bool>(const In
   if (right_expression.type == ExpressionType::List) {
     const auto& array_expression = static_cast<const ListExpression&>(right_expression);
 
+    DebugAssert(!array_expression.elements().empty(), "IN clauses with an empty list are invalid");
+
+    if (left_expression.data_type() == DataType::Null) {
+      // `NULL IN ...` is NULL
+      return std::make_shared<ExpressionResult<ExpressionEvaluator::Bool>>(std::vector<ExpressionEvaluator::Bool>{0},
+                                                                           std::vector<bool>{1});
+    }
+
     /**
-     * To keep the code simple for now, transform the InExpression like this:
-     * "a IN (x, y, z)"   ---->   "a = x OR a = y OR a = z"
-     *
-     * But first, out of array_expression.elements(), pick those expressions whose type can be compared with
+     * Out of array_expression.elements(), pick those expressions whose type can be compared with
      * in_expression.value() so we're not getting "Can't compare Int and String" when doing something crazy like
      * "5 IN (6, 5, "Hello")
      */
     const auto left_is_string = left_expression.data_type() == DataType::String;
     std::vector<std::shared_ptr<AbstractExpression>> type_compatible_elements;
-    for (const auto& element : array_expression.elements()) {
-      if ((element->data_type() == DataType::String) == left_is_string) {
-        type_compatible_elements.emplace_back(element);
+    bool all_elements_are_values_of_left_type = true;
+    resolve_data_type(left_expression.data_type(), [&](const auto left_data_type_t) {
+      using LeftDataType = typename decltype(left_data_type_t)::type;
+
+      for (const auto& element : array_expression.elements()) {
+        if ((element->data_type() == DataType::String) == left_is_string) {
+          type_compatible_elements.emplace_back(element);
+        }
+
+        if (element->type != ExpressionType::Value) {
+          all_elements_are_values_of_left_type = false;
+        } else {
+          const auto& value_expression = std::static_pointer_cast<ValueExpression>(element);
+          if (value_expression->value.type() != typeid(LeftDataType)) all_elements_are_values_of_left_type = false;
+        }
       }
-    }
+    });
 
     if (type_compatible_elements.empty()) {
-      // `5 IN ()` is FALSE as is `NULL IN ()`
+      // `5 IN ()` is FALSE
       return std::make_shared<ExpressionResult<ExpressionEvaluator::Bool>>(std::vector<ExpressionEvaluator::Bool>{0});
     }
 
+    // If all elements of the list are simple values (e.g., `IN (1, 2, 3)`), iterate over the column and compare the
+    // values. Otherwise, translate the IN clause to a series of ORs:
```

Comment:
this is missing the key point: If both the left argument and all elements in the list are the same type, then we can do a **binary search** through the list, which is faster and worth the hassle to have a separate implementation for.

Reply:
The binary search is not the main contribution. Even when doing a linear search, we get almost the same performance. The binary search is just because of its algorithmic beauty (and for reeeeally long lists, as they might come from ORMs).

Still, I will update the comment.

Reply:
Yeah, as long as there is some justification for the code path I'm happy.